packages/seacas/libraries/ioss/src/Ioss_DynamicTopology.C

```diff
@@ -36,14 +36,463 @@
 #include <fmt/format.h>
 #include <fmt/ostream.h>
 #include <string>
-
+#include <functional>
 #include <iostream>
 #include <sstream>
 #include <iomanip>
 #include <assert.h>
 
 #include "Ioss_ParallelUtils.h"
 
+namespace {
+
+struct DatabaseState
+{
+  DatabaseState(Ioss::DatabaseIO* db)
+  {
+    if(!db->supports_change_set()) {
+      changeSet = db->get_filename();
+    }
+  }
+
+  std::string changeSet{"/"};
+  int state{-1};
+  double time{-std::numeric_limits<double>::max()};
+};
+
+using StateLocatorCompare = std::function<bool(double, double)>;
+
+void locate_state_impl(Ioss::DatabaseIO *db, double targetTime,
+                       StateLocatorCompare comparator, DatabaseState& loc)
+{
+  std::vector<double> timesteps = db->get_db_step_times();
+  int stepCount = timesteps.size();
+
+  double minTimeDiff = loc.state < 0 ? std::numeric_limits<double>::max() : std::fabs(loc.time - targetTime);
+
+  for(int istep = 1; istep <= stepCount; istep++) {
+    double stateTime = timesteps[istep-1];
+    double stepTimeDiff = std::fabs(stateTime - targetTime);
+    if(comparator(stepTimeDiff, minTimeDiff)) {
+      minTimeDiff = stepTimeDiff;
+      loc.time  = stateTime;
+      loc.state = istep;
+      loc.changeSet = db->supports_change_set() ? db->get_change_set_name() : db->get_filename();
+    }
+  }
+}
+
+void locate_state(Ioss::DatabaseIO *db, double targetTime, DatabaseState& loc)
+{
+  // Get state count and all states...
+  std::vector<double> timesteps = db->get_db_step_times();
+  int stepCount = timesteps.size();
+
+  if(targetTime < 0.0) {
```

Comment:
This is much clearer than previous code.

Reply:
The power of refactoring 😸